src/Type/ArrayType.php

```diff
@@ -288,6 +289,30 @@ public static function castToArrayKeyType(Type $offsetType): Type
 		return new UnionType([new IntegerType(), new StringType()]);
 	}
 
+	public function inferTemplateTypes(Type $receivedType): TemplateTypeMap
+	{
+		if ($receivedType instanceof UnionType || $receivedType instanceof IntersectionType) {
+			return $receivedType->inferTemplateTypesOn($this);
+		}
+
+		if (
+			$receivedType instanceof ArrayType
+			&& $this->getKeyType()->accepts($receivedType->getKeyType(), true)->yes()
```

Comment:
`isSuperTypeOf` is a better fit that should be used in most places; `accepts` is just for user-facing rules and accomodates some weird PHP quirks (like objects with __toString() passed to strings, ints passed to floats etc.).

Reply:
Here I'm really checking if some parameter would accept from value; I though it would be technically and semantically correct to do this.

Now that you are saying this, I notice that ParametersAcceptorSelector uses isSuperTypeOf(), while FunctionCallParametersCheck uses accepts(). Is this intentional ? Do we expect accepts() and isSuperTypeOf() to behave differently ?

Reply:
Yes, that’s intentional. FunctionCallParametersCheck is just for rules. isSuperTypeOf is more correct when looking at the type hierarchy